clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -2473,9 +2475,116 @@ class SyclKernelIntHeaderCreator : public SyclKernelFieldHandler {
 
 } // namespace
 
+class SYCLKernelNameTypeVisitor
+    : public TypeVisitor<SYCLKernelNameTypeVisitor>,
+      public ConstTemplateArgumentVisitor<SYCLKernelNameTypeVisitor> {
+  Sema &S;
+  SourceLocation Loc;
+  using InnerTypeVisitor = TypeVisitor<SYCLKernelNameTypeVisitor>;
+  using InnerTAVisitor =
+      ConstTemplateArgumentVisitor<SYCLKernelNameTypeVisitor>;
+
+public:
+  SYCLKernelNameTypeVisitor(Sema &S, SourceLocation Loc) : S(S), Loc(Loc) {}
+
+  void Visit(QualType T) {
+    if (T.isNull())
+      return;
+    const CXXRecordDecl *RD = T->getAsCXXRecordDecl();
+    if (!RD)
+      return;
+    // If KernelNameType has template args visit each template arg via
+    // ConstTemplateArgumentVisitor
+    if (const auto *TSD = dyn_cast<ClassTemplateSpecializationDecl>(RD)) {
+      const TemplateArgumentList &Args = TSD->getTemplateArgs();
+      for (unsigned I = 0; I < Args.size(); I++) {
+        const TemplateArgument &TemplateArg = Args[I];
+        Visit(TemplateArg);
+      }
+    } else {
+      InnerTypeVisitor::Visit(T.getTypePtr());
+    }
+  }
+
+  void Visit(const TemplateArgument &TA) { InnerTAVisitor::Visit(TA); }
+
+  void VisitEnumType(const EnumType *T) {
+    const EnumDecl *ED = T->getDecl();
+    if (!ED->isScoped() && !ED->isFixed()) {
+      S.Diag(Loc, diag::err_sycl_kernel_incorrectly_named) << 2;
+      S.Diag(ED->getSourceRange().getBegin(), diag::note_entity_declared_at)
+          << ED;
+      return;
+    }
+  }
+
+  void VisitRecordType(const RecordType *T) {
+    return VisitTagDecl(T->getDecl());
+  }
+
+  void VisitTagDecl(const TagDecl *Tag) {
+    bool UnnamedLambdaEnabled =
+        S.getASTContext().getLangOpts().SYCLUnnamedLambda;
+    if (Tag && !Tag->getDeclContext()->isTranslationUnit() &&
+        !dyn_cast_or_null<NamespaceDecl>(Tag->getDeclContext()) &&
```

Comment:
Suggest use `isa` here since you don't use result of `dyn_cast_or_null`
```suggestion
        !isa<NamespaceDecl>(Tag->getDeclContext()) &&
```